rllib/agents/trainer_template.py

```diff
@@ -139,27 +139,34 @@ def _init(self, config: TrainerConfigDict,
                 policy_class=self._policy_class,
                 config=config,
                 num_workers=self.config["num_workers"])
-            # If execution plan is not provided (None), the Trainer will use
-            # it's already existing default `execution_plan()` static method
-            # instead.
-            if execution_plan is not None:
-                self.execution_plan = execution_plan
+
             self.train_exec_impl = self.execution_plan(
                 self.workers, config, **self._kwargs_for_execution_plan())
 
             if after_init:
                 after_init(self)
 
-        @staticmethod
         @override(Trainer)
-        def _validate_config(config: PartialTrainerConfigDict,
-                             trainer_obj_or_none: Optional["Trainer"] = None):
+        def validate_config(self, config: PartialTrainerConfigDict):
             # Call super (Trainer) validation method first.
-            Trainer._validate_config(config, trainer_obj_or_none)
+            Trainer.validate_config(self, config)
             # Then call user defined one, if any.
             if validate_config is not None:
                 validate_config(config)
 
+        @staticmethod
+        @override(Trainer)
+        def execution_plan(workers, config, **kwargs):
+            # `execution_plan` is provided, use it inside
+            # `self.execution_plan()`.
+            if execution_plan is not None:
+                return execution_plan(workers, config, **kwargs)
```

Comment:
wait, I don't get why we need this all of a sudden. is this really related to APPO trainer?
where does this execution_plan come from?

Reply:
Just keeping things backward compatible (with the soon-to-be-deprecated `build_trainer` function).

New:
- Trainer now has a method: `execution_plan` with the default implementation (the one that e.g. PGTrainer uses).
- Users can override it.
- However, it is still a static method, such that backward compatibility with `build_trainer` will not break.
- It was necessary to introduce this in this PR as APPO is the first agent that we convert to sub-class that brings its own exec plan logic (PGTrainer used the default one; all other Trainers are still using `build_trainer`).

Reply:
I am still a bit confused about how this gets called.
because won't the first condition always be true? like we are in execution_plan itself, so execution_plan must be there, non-None?